### PR TITLE
Add command to rotate e2e credentials

### PR DESCRIPTION
**Pull request by hawkowl**
### Which issue this PR addresses:

Fixes ARO-6742

### What this PR does / why we need it:

Generates/rotates passwords for the SPNs

### Test plan for issue:
Tested in dev, don't have permissions for INT, will require a TL to do that

### Is there any documentation that needs to be updated for this PR?

SOPs for rotation

### How do you know this will function as expected in production? 

I don't, but it should be re-runnable...

### CODE REVIEW

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+			return err
+		}
+		return nil
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(generateServicePrincipalsCmd)
+
+	generateServicePrincipalsCmd.Flags().String("keyvault", "aro-e2e-principals", "keyvault to use")
```
Comment (mociarain) [inline hack/e2e/cmd/generateServicePrincipals.go:42]: I like this pattern. Maybe I'm extrapolating too much (I find the cobra docs hard to parse) but if I understand it correctly we could replace a lot of the `go run ......` stuff with this and "proper" CLI commands?

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	"github.com/Azure/go-autorest/autorest/date"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/jongio/azidext/go/azidext"
+	"github.com/spf13/cobra"
+
+	"github.com/Azure/ARO-RP/pkg/env"
+	msgraph_apps "github.com/Azure/ARO-RP/pkg/util/graph/graphsdk/applications"
+	msgraph_models "github.com/Azure/ARO-RP/pkg/util/graph/graphsdk/models"
+	"github.com/Azure/ARO-RP/pkg/util/graph/graphsdk/serviceprincipals"
+	"github.com/Azure/ARO-RP/pkg/util/keyvault"
```
Comment (bitoku) [inline hack/e2e/cmd/generateServicePrincipals.go:21]: We may want to use the new keyvault track2 SDK instead of the old one since the keyvault SDK track2 is merged https://github.com/Azure/ARO-RP/pull/3275/files

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+	keyvaultURI := keyvault.URI(_env, keyvaultParam, "")
+	keyvaultManager := keyvault.NewManager(cred, keyvaultURI)
+
+	graphClient, err := _env.Environment().NewGraphServiceClient(azcred)
+	if err != nil {
+		return err
+	}
+
+	for i := 1; i == spnCountParam; i++ {
```
Comment (AldoFusterTurpin) [inline hack/e2e/cmd/generateServicePrincipals.go:86]: Shouldn't this be
``` 
for i := 0; i < spnCountParam; i++ {
```
but not "==" ? 
Maybe I am missing something.

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+			return fmt.Errorf("%d applications found for spn '%s'", len(apps), spnName)
+		}
+
+		log.Infof("found IDs for '%s': ObjectID='%s', AppID='%s'", spnName, spnObjectID, spnAppID)
+
+		log.Info("Ensuring SPN ID in keyvault...")
+
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-app-id", spnName), azkeyvault.SecretSetParameters{
+			Value: to.StringPtr(spnAppID),
+		})
```
Comment (AldoFusterTurpin) [inline hack/e2e/cmd/generateServicePrincipals.go:123]: IMHO extracting into a variable inline structs that are > 1 line long is a good thing (talking about azkeyvault.SecretSetParameters{)

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		if err != nil {
+			return err
+		}
+
+		if len(apps[0].GetPasswordCredentials()) > 0 {
```
Comment (AldoFusterTurpin) [inline hack/e2e/cmd/generateServicePrincipals.go:128]: We are using _apps[0].GetPasswordCredentials()_ in 3 different places, maybe we can extract it to a variable, more performant and easier to read.

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+			return err
+		}
+
+		log.Info("setting password credential for SPN in keyvault")
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-secret-value", spnName), azkeyvault.SecretSetParameters{
+			Value: resp.GetSecretText(),
+			SecretAttributes: &azkeyvault.SecretAttributes{
+				Expires: &expiryTime,
+			},
+		})
```
Comment (AldoFusterTurpin) [inline hack/e2e/cmd/generateServicePrincipals.go:167]: Same comment as before, I think it would be great to extract _azkeyvault.SecretSetParameters_ into a variable as this is not easy to read (at least for me).

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+		if err != nil {
+			return err
+		}
+
+		log.Info("checking for service principal")
+
+		filter = fmt.Sprintf("appId eq '%s'", spnAppID)
+		spResult, err := graphClient.ServicePrincipals().Get(ctx, &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{
```
Comment (AldoFusterTurpin) [inline hack/e2e/cmd/generateServicePrincipals.go:175]: Same as before for _serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration_

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+			spID = *spAddResult.GetId()
+		case 1:
+			spID = *sps[0].GetId()
+			log.Infof("found SP ID %s", spID)
+		default:
+			return fmt.Errorf("%d sps found for spn '%s'", len(sps), spnName)
+		}
+
+		log.Info("setting service principal ID in keyvault")
+		err = keyvaultManager.SetSecret(ctx, fmt.Sprintf("%s-sp-id", spnName), azkeyvault.SecretSetParameters{
```
Comment (AldoFusterTurpin) [inline hack/e2e/cmd/generateServicePrincipals.go:207]: Same as before for _azkeyvault.SecretSetParameters_. Even here is not that important, I still see value in extracting it.

File: hack/e2e/cmd/generateServicePrincipals.go [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+		err := generateKeys(cmd)
+		if err != nil {
+			cmd.PrintErr(err)
+			return err
+		}
+		return nil
+	},
+}
+
+func init() {
```
Comment (AldoFusterTurpin) [inline hack/e2e/cmd/generateServicePrincipals.go:39]: Sadly, Cobra forces the developer to use the init() function to register the commands/subcommands and "fun" problems can arise with the usage of init() in Go. But what can we do ...? 😆

**Review by AldoFusterTurpin — COMMENTED** [review]
Thank you for this @hawkowl. I just left some optional suggestions and a question about a for loop.

**Comment by hawkowl** [comment]
I don't think we need this anymore.
